### PR TITLE
[SPARK-18810][SPARKR] SparkR install.spark does not work for RCs, snapshots

### PR DESCRIPTION
**Pull request by felixcheung**
## What changes were proposed in this pull request?

Support overriding the download url (include version directory) in an environment variable, `SPARKR_RELEASE_DOWNLOAD_URL`

## How was this patch tested?

unit test, manually testing
- snapshot build url
  - download when spark jar not cached
  - when spark jar is cached
- RC build url
  - download when spark jar not cached
  - when spark jar is cached
- multiple cached spark versions
- starting with sparkR shell

To use this,
```
SPARKR_RELEASE_DOWNLOAD_URL=http://this_is_the_url_to_spark_release_tgz R
```
then in R,
```
library(SparkR) # or specify lib.loc
sparkR.session()
```

### CODE REVIEW

**Comment by felixcheung** [comment]
Jenkins, retest this please

**Comment by felixcheung** [comment]
appveyer passed before.

**Comment by felixcheung** [comment]
@shivaram all tests passed.

File: R/pkg/R/install.R [hunk 1]
```diff
@@ -104,7 +113,12 @@ install.spark <- function(hadoopVersion = "2.7", mirrorUrl = NULL,
   if (tarExists && !overwrite) {
     message("tar file found.")
   } else {
-    robustDownloadTar(mirrorUrl, version, hadoopVersion, packageName, packageLocalPath)
+    if (releaseUrl != "") {
+      message("Downloading from alternate URL:\n- ", releaseUrl)
+      downloadUrl(releaseUrl, packageLocalPath, paste0("Fetch failed from ", mirrorUrl))
```
Comment (shivaram) [inline R/pkg/R/install.R]: this should be `releaseUrl` instead of `mirrorUrl` in the `paste0` ?

File: R/pkg/R/utils.R [hunk 2]
```diff
@@ -851,3 +851,12 @@ rbindRaws <- function(inputData){
   out[!rawcolumns] <- lapply(out[!rawcolumns], unlist)
   out
 }
+
+# Get basename without extension from URL
+basenameSansExtFromUrl <- function(url) {
```
Comment (shivaram) [inline R/pkg/R/utils.R]: can we use file_path_sans_ext [1] for removing the extension ? I worry we might publish it as `.tar.gz` someday and then removing just the last `.` will be insufficient

[1] https://stat.ethz.ch/R-manual/R-patched/library/tools/html/fileutils.html

Comment (felixcheung) [inline R/pkg/R/utils.R]: My concern was to bring in another dependencies just for this (it's in the tools)

The regex was in fact copy-paste from file_path_sans_ext (hence the name) except for the compression part which is what you are referring to.
I could copy that over as well. Would you prefer `compression` be TRUE (default is FALSE) to remove `.gz`?

```
> library(tools)
> file_path_sans_ext
function (x, compression = FALSE)
{
    if (compression)
        x <- sub("[.](gz|bz2|xz)$", "", x)
    sub("([^.]+)\\.[[:alnum:]]+$", "\\1", x)
}
```

Comment (shivaram) [inline R/pkg/R/utils.R]: Got it - yeah it sounds good to remove compression and just copy the two lines of code.

**Review by shivaram — COMMENTED** [review]
LGTM. Thanks @felixcheung - I'll just test this locally once and then merge

**Comment by shivaram** [comment]
Tested locally - Works fine. Merging into master, branch-2.1

**Pull request closed without merging**